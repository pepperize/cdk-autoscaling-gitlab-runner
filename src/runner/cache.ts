import { Duration, RemovalPolicy, Stack } from "aws-cdk-lib";
import { BlockPublicAccess, Bucket, BucketEncryption, IBucket } from "aws-cdk-lib/aws-s3";
import { Construct } from "constructs";

export interface CacheProps {
  /**
   * The infix of the physical cache bucket name.
   *
   * @default "runner-cache"
   */
  readonly bucketName?: string;
  /**
   * The number of days after which the created cache objects are deleted from S3.
   * @default 30 days
   */
  readonly expiration?: Duration;
}

/**
 * A GitLab Runner cache consisting of an Amazon S3 bucket.
 *
 * The bucket is encrypted with a KMS managed master key, it has public access blocked and will be cleared and deleted on CFN stack deletion.
 */
export class Cache extends Construct {
  readonly bucket: IBucket;

  constructor(scope: Stack, id: string, props: CacheProps = {}) {
    super(scope, id);

    const bucketName = props.bucketName || "runner-cache";
    const uniqueCacheBucketName = `${scope.stackName}-${bucketName}-${scope.account}-${scope.region}`
      .slice(-63)
      .toLocaleLowerCase();

    /* Enabled if not 0. If 0 - cache doesn't expire. If undefined - expiration sets to expire in 30 days */
    const expiration = props.expiration ?? Duration.days(30);
    const lifeCycleRuleEnabled = expiration.toDays() !== 0;

    this.bucket = new Bucket(scope, "CacheBucket", {
      bucketName: uniqueCacheBucketName,
      lifecycleRules: [
        {
          enabled: lifeCycleRuleEnabled,
          expiration: expiration,
        },
      ],
      encryption: BucketEncryption.KMS_MANAGED,
      blockPublicAccess: BlockPublicAccess.BLOCK_ALL,
      autoDeleteObjects: true,
      removalPolicy: RemovalPolicy.DESTROY,
    });
  }
}
