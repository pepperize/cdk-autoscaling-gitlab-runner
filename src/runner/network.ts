import { Annotations, Stack } from "aws-cdk-lib";
import { ISubnet, IVpc, SubnetSelection, SubnetType, Vpc } from "aws-cdk-lib/aws-ec2";
import { Construct } from "constructs";

export interface NetworkProps {
  /**
   * If no existing VPC is provided, a default Vpc will be created.
   */
  readonly vpc?: IVpc;

  /**
   * The GitLab Runner's subnets. It should be either public or private. If more then subnet is selected, then the first found (private) subnet will be used.
   * @see https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-ec2.SubnetSelection.html
   *
   * A network is considered private, if
   *  - tagged by 'aws-cdk:subnet-type'
   *  - doesn't route to an Internet Gateway (not public)
   *  - has an Nat Gateway
   */
  readonly subnetSelection?: SubnetSelection;
}

/**
 * Network settings for the manager and runners
 *
 *  All EC2 instances should belong to the same subnet, availability zone and vpc.
 */
export class Network extends Construct {
  readonly vpc: IVpc;
  readonly availabilityZone: string;
  readonly subnet: ISubnet;

  constructor(scope: Stack, id: string, props?: NetworkProps) {
    super(scope, id);

    this.vpc =
      props?.vpc ??
      new Vpc(scope, `Vpc`, {
        maxAzs: 1,
      });

    this.subnet = this.findSubnet(this.vpc, props?.subnetSelection);

    this.availabilityZone = this.subnet.availabilityZone;

    if (!this.hasPrivateSubnet(this.vpc)) {
      Annotations.of(this).addWarning(`No private network found in ${this.vpc.vpcId}, using public addresses.`);
    }
  }

  private hasPrivateSubnet(vpc: IVpc): boolean {
    return !!vpc.privateSubnets.length;
  }

  /**
   * Returns the first private or public subnet. Optionally filters by AZ.
   *
   * @exception Throws an error if no private or public is found.
   */
  private findSubnet(vpc: IVpc, subnetSelection?: SubnetSelection): ISubnet {
    const selectedSubnets = vpc.selectSubnets(
      subnetSelection || {
        subnetType: this.hasPrivateSubnet(vpc) ? SubnetType.PRIVATE_WITH_NAT : SubnetType.PUBLIC,
        availabilityZones: vpc.availabilityZones,
      }
    );

    const subnet = selectedSubnets.subnets.find(() => true);

    if (!subnet) {
      throw new Error(`Neither a private nor a public subnet is found in ${vpc.vpcId}`);
    }

    return subnet;
  }
}
