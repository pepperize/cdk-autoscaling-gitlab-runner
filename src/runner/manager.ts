import { SecurityGroup } from "@pepperize/cdk-security-group";
import { Stack } from "aws-cdk-lib";
import {
  AmazonLinuxCpuType,
  AmazonLinuxEdition,
  AmazonLinuxGeneration,
  AmazonLinuxStorage,
  AmazonLinuxVirt,
  CloudFormationInit,
  IMachineImage,
  InitCommand,
  InitConfig,
  InitFile,
  InitPackage,
  InitService,
  InitServiceRestartHandle,
  InstanceClass,
  InstanceSize,
  InstanceType,
  MachineImage,
  UserData,
} from "aws-cdk-lib/aws-ec2";
import { IRole, ManagedPolicy, PolicyDocument, Role, ServicePrincipal } from "aws-cdk-lib/aws-iam";
import { IBucket } from "aws-cdk-lib/aws-s3";
import { Construct } from "constructs";
import { GitlabRunnerAutoscalingJobRunner } from "./job-runner";
import { Network } from "./network";
import { ConfigurationMapper, GlobalConfiguration, RunnerConfiguration } from "../runner-configuration";

export interface GitlabRunnerAutoscalingManagerBaseProps {
  /**
   * An Amazon Machine Image ID for the Manager EC2 instance. If empty the latest Amazon 2 Image will be looked up.
   *
   * Should be RHEL flavor like Amazon Linux 2 with yum available for instance initialization.
   *
   * @see https://cloudinit.readthedocs.io/en/latest/
   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/cfn-init.html
   */
  readonly machineImage?: IMachineImage;

  /**
   * Instance type for manager EC2 instance. It's a combination of a class and size.
   * @default InstanceType.of(InstanceClass.T3, InstanceSize.NANO)
   */
  readonly instanceType?: InstanceType;

  /**
   * A set of security credentials that you use to prove your identity when connecting to an Amazon EC2 instance. You won't be able to ssh into an instance without the Key Pair.
   */
  readonly keyPairName?: string;
}

export interface GitlabRunnerAutoscalingManagerProps extends GitlabRunnerAutoscalingManagerBaseProps {
  readonly globalConfiguration?: GlobalConfiguration;
  readonly runners: GitlabRunnerAutoscalingJobRunner[];
  readonly network: Network;
  readonly cacheBucket: IBucket;
  readonly role?: IRole;
  readonly runnersSecurityGroup: SecurityGroup;
}

const DEFAULT_SSH_KEY_PATH = "/etc/gitlab-runner/ssh-custom";

/**
 * Settings for the manager (coordinator)
 *
 *  Manager coordinates the placement of runner (job executor) instances
 */
export class GitlabRunnerAutoscalingManager extends Construct {
  readonly machineImage: IMachineImage;
  readonly instanceType: InstanceType;
  readonly keyPairName?: string;
  readonly runners: GitlabRunnerAutoscalingJobRunner[];
  readonly network: Network;
  readonly runnersSecurityGroupName: string;
  readonly role: IRole;
  readonly initConfig: CloudFormationInit;
  readonly userData: UserData;
  readonly cacheBucket: IBucket;
  readonly globalConfiguration: GlobalConfiguration;

  constructor(scope: Construct, id: string, props: GitlabRunnerAutoscalingManagerProps) {
    super(scope, id);
    this.globalConfiguration =
      props.globalConfiguration ||
      ({
        concurrent: 10,
        checkInterval: 0,
        logFormat: "runner",
        logLevel: "info",
      } as GlobalConfiguration);
    this.machineImage =
      props.machineImage ??
      MachineImage.latestAmazonLinux({
        generation: AmazonLinuxGeneration.AMAZON_LINUX_2,
        edition: AmazonLinuxEdition.STANDARD,
        virtualization: AmazonLinuxVirt.HVM,
        storage: AmazonLinuxStorage.EBS,
        cpuType: AmazonLinuxCpuType.X86_64,
      });
    this.instanceType = props.instanceType ?? InstanceType.of(InstanceClass.T3, InstanceSize.NANO);
    this.keyPairName = props.keyPairName;
    this.runners = props.runners;
    this.network = props.network;
    this.cacheBucket = props.cacheBucket;
    this.runnersSecurityGroupName = props.runnersSecurityGroup.securityGroupName;

    this.role =
      props.role ||
      new Role(scope, "ManagerRole", {
        assumedBy: new ServicePrincipal("ec2.amazonaws.com", {}),
        managedPolicies: [ManagedPolicy.fromAwsManagedPolicyName("AmazonSSMManagedInstanceCore")],
        inlinePolicies: {
          Cache: PolicyDocument.fromJson({
            Version: "2012-10-17",
            Statement: [
              {
                Effect: "Allow",
                Action: ["s3:ListObjects*", "s3:GetObject*", "s3:DeleteObject*", "s3:PutObject*"],
                Resource: [`${this.cacheBucket.bucketArn}/*`],
              },
              {
                Effect: "Allow",
                Action: ["s3:ListBucket"],
                Resource: [`${this.cacheBucket.bucketArn}`],
              },
            ],
          }),
          Runners: PolicyDocument.fromJson({
            Version: "2012-10-17",
            Statement: [
              {
                Effect: "Allow",
                Action: ["ec2:CreateKeyPair", "ec2:DeleteKeyPair", "ec2:ImportKeyPair", "ec2:Describe*"],
                Resource: ["*"],
              },
              {
                Effect: "Allow",
                Action: ["ec2:CreateTags", "ssm:UpdateInstanceInformation"],
                Resource: ["*"],
                Condition: {
                  StringLike: {
                    "aws:RequestTag/Name": "*gitlab-runner-*",
                  },
                  "ForAllValues:StringEquals": {
                    "aws:TagKeys": ["Name"],
                  },
                },
              },
              {
                Effect: "Allow",
                Action: ["ec2:RequestSpotInstances", "ec2:CancelSpotInstanceRequests"],
                Resource: ["*"],
                Condition: {
                  StringEqualsIfExists: {
                    "ec2:Region": `${Stack.of(this).region}`,
                  },
                  ArnEqualsIfExists: {
                    "ec2:Vpc": `arn:aws:ec2:${Stack.of(this).region}:${Stack.of(this).account}:vpc/${
                      this.network.vpc.vpcId
                    }`,
                  },
                },
              },
              {
                Effect: "Allow",
                Action: ["ec2:RunInstances"],
                Resource: ["*"],
                Condition: {
                  "ForAllValues:StringEquals": {
                    "aws:TagKeys": ["InstanceProfile"],
                    "ec2:InstanceType": (this.runners || []).map((runner) => {
                      const runnersInstanceType =
                        (this.runners && runner.instanceType) || InstanceType.of(InstanceClass.T3, InstanceSize.MICRO);
                      return runnersInstanceType.toString();
                    }),
                  },
                },
              },
              {
                Effect: "Allow",
                Action: ["ec2:TerminateInstances", "ec2:StopInstances", "ec2:StartInstances", "ec2:RebootInstances"],
                Resource: ["*"],
                Condition: {
                  StringLike: {
                    "ec2:ResourceTag/Name": "*gitlab-runner-*",
                  },
                },
              },
              {
                Effect: "Allow",
                Action: ["iam:PassRole"],
                Resource: ["*"],
                Condition: {
                  "ForAllValues:StringEquals": {
                    "aws:TagKeys": ["RunnersRole"],
                  },
                },
              },
            ],
          }),
        },
      });

    this.userData = UserData.forLinux({});
    this.userData.addCommands(
      `yum update -y aws-cfn-bootstrap` // !/bin/bash -xe
    );

    // https://github.com/awslabs/amazon-ecr-credential-helper
    const userDataRunners = UserData.forLinux({});
    userDataRunners.addCommands(
      `[ ! -z "$(which apt-get)" ] && apt-get install -y amazon-ecr-credential-helper`,
      `[ ! -z "$(which yum)" ] && yum install -y amazon-ecr-credential-helper`
    );

    const gitlabRunnerConfigRestartHandle = new InitServiceRestartHandle();
    gitlabRunnerConfigRestartHandle._addFile("/etc/gitlab-runner/config.toml");

    const rsyslogConfigRestartHandle = new InitServiceRestartHandle();
    rsyslogConfigRestartHandle._addFile("/etc/rsyslog.d/25-gitlab-runner.conf");

    /**
     * Config set keys
     * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-init.html#aws-resource-init-configsets
     */
    const REPOSITORIES = "repositories";
    const PACKAGES = "packages";
    const CONFIG = "config";
    const RESTART = "restart";

    this.initConfig = CloudFormationInit.fromConfigSets({
      configSets: {
        default: [REPOSITORIES, PACKAGES, CONFIG, RESTART],
      },
      configs: {
        [REPOSITORIES]: new InitConfig([
          InitCommand.shellCommand(
            "curl -L https://packages.gitlab.com/install/repositories/runner/gitlab-runner/script.rpm.sh | bash",
            { key: "10-gitlab-runner" }
          ),
        ]),
        [PACKAGES]: new InitConfig([
          InitPackage.yum("docker"),
          InitPackage.yum("gitlab-runner"),
          InitPackage.yum("tzdata"),
          InitPackage.yum("jq"),
          InitCommand.shellCommand(
            "curl -L https://gitlab-docker-machine-downloads.s3.amazonaws.com/v0.16.2-gitlab.19/docker-machine-`uname -s`-`uname -m` > /tmp/docker-machine && install /tmp/docker-machine /usr/bin/docker-machine",
            //"curl -L https://github.com/docker/machine/releases/download/v0.16.2/docker-machine-`uname -s`-`uname -m` > /tmp/docker-machine && install /tmp/docker-machine /usr/bin/docker-machine",
            { key: "10-docker-machine" }
          ),
          InitCommand.shellCommand("gitlab-runner start", {
            key: "20-gitlab-runner-start",
          }),
        ]),
        [CONFIG]: new InitConfig([
          InitFile.fromString(
            "/etc/gitlab-runner/config.toml",
            ConfigurationMapper.withDefaults({
              globalConfiguration: this.globalConfiguration,
              runnersConfiguration: this.runners.map((runner) => {
                const configuration = runner.configuration;
                return {
                  ...configuration,
                  machine: {
                    ...configuration.machine,
                    machineOptions: {
                      sshKeypath:
                        // drivers/amazonec2/amazonec2.go SSHPrivateKeyPath
                        runner.keyPair
                          ? `${DEFAULT_SSH_KEY_PATH}/${configuration.machine?.machineOptions?.keypairName}`
                          : "",
                      ...configuration.machine?.machineOptions,
                      instanceType: runner.instanceType.toString(),
                      ami: runner.machineImage.getImage(scope).imageId,
                      region: Stack.of(this).region,
                      vpcId: this.network.vpc.vpcId,
                      zone: this.network.availabilityZone.slice(-1),
                      subnetId: this.network.subnet.subnetId,
                      securityGroup: this.runnersSecurityGroupName,
                      privateAddressOnly:
                        configuration.machine?.machineOptions?.privateAddressOnly ?? this.network.hasPrivateSubnets(),
                      usePrivateAddress: configuration.machine?.machineOptions?.usePrivateAddress ?? true,
                      iamInstanceProfile: runner.instanceProfile.ref,
                      userdata: "/etc/gitlab-runner/user_data_runners",
                    },
                  },
                  cache: {
                    s3: {
                      serverAddress: `s3.${Stack.of(this).urlSuffix}`,
                      bucketName: `${this.cacheBucket.bucketName}`,
                      bucketLocation: `${Stack.of(this).region}`,
                      authenticationType: "iam",
                    },
                  },
                } as RunnerConfiguration;
              }),
            }).toToml(),
            {
              owner: "gitlab-runner",
              group: "gitlab-runner",
              mode: "000600",
            }
          ),
          InitFile.fromString(
            "/etc/rsyslog.d/25-gitlab-runner.conf",
            `:programname, isequal, "gitlab-runner" /var/log/gitlab-runner.log`,
            {
              owner: "root",
              group: "root",
              mode: "000644",
            }
          ),
          InitService.enable("gitlab-runner", {
            ensureRunning: true,
            enabled: true,
            serviceRestartHandle: gitlabRunnerConfigRestartHandle,
          }),
          InitService.enable("rsyslog", {
            ensureRunning: true,
            enabled: true,
            serviceRestartHandle: rsyslogConfigRestartHandle,
          }),
          InitCommand.shellCommand(
            // Download custom EC2 key pair for manager <> runner ssh connect
            this.runners
              .map((runner) => {
                if (!runner.keyPair) {
                  return "";
                }

                runner.keyPair.grantRead(this.role);

                const region = Stack.of(this).region;
                const secretArn = runner.keyPair.secretArn;
                const keyPairName = runner.configuration.machine!.machineOptions!.keypairName;
                const sshKeyPath = runner.configuration.machine!.machineOptions!.sshKeypath ?? DEFAULT_SSH_KEY_PATH;

                return [
                  `mkdir -p ${sshKeyPath};`,
                  `echo $(aws secretsmanager get-secret-value --region ${region} --secret-id ${secretArn} --query SecretString --output text) | jq -r '."${keyPairName}"' > ${sshKeyPath}/${keyPairName};`,
                  `echo $(aws secretsmanager get-secret-value --region ${region} --secret-id ${secretArn} --query SecretString --output text) | jq -r '."${keyPairName}.pub"' > ${sshKeyPath}/${keyPairName}.pub;`,
                ].join("\n");
              })
              .filter((s) => s.length > 0)
              .join("\n"),
            {
              key: "999-retrieve-ec2-key-pair",
            }
          ),
          InitFile.fromString("/etc/gitlab-runner/user_data_runners", userDataRunners.render(), {
            owner: "gitlab-runner",
            group: "gitlab-runner",
            mode: "000600",
          }),
        ]),
        [RESTART]: new InitConfig([
          InitCommand.shellCommand("gitlab-runner restart", {
            key: "10-gitlab-runner-restart",
          }),
        ]),
      },
    });
  }
}
