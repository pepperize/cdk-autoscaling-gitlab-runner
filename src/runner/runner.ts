import { SecurityGroup } from "@pepperize/cdk-security-group";
import { Duration, Stack } from "aws-cdk-lib";
import { AutoScalingGroup, Signals } from "aws-cdk-lib/aws-autoscaling";
import { Port } from "aws-cdk-lib/aws-ec2";
import { IBucket } from "aws-cdk-lib/aws-s3";
import { Construct } from "constructs";
import { Cache, CacheProps } from "./cache";
import { GitlabRunnerAutoscalingJobRunner, GitlabRunnerAutoscalingJobRunnerProps } from "./job-runner";
import { GitlabRunnerAutoscalingManager, GitlabRunnerAutoscalingManagerBaseProps } from "./manager";
import { Network, NetworkProps } from "./network";
import { GlobalConfiguration, LogFormat, LogLevel } from "../runner-configuration";

/**
 * This is a AWS CDK Construct that may be used to deploy a GitLab runner with Docker executor and auto-scaling.
 *
 * @remarks
 * The `cdk-gitlab-runner` defines the {@link GitlabRunnerAutoscalingProps} interface and {@link GitlabRunnerAutoscaling} construct class,
 * which are used to provision a the runner.
 *
 * @packageDocumentation
 */

/**
 * Properties of the Gitlab Runner. You have to provide at least the GitLab's Runner's authentication token.
 */
export interface GitlabRunnerAutoscalingProps extends GlobalConfiguration {
  readonly cache?: GitlabRunnerAutoscalingCacheProps;

  /**
   * The network configuration for the Runner. If not set, the defaults will be used.
   * @link NetworkProps
   */
  readonly network?: NetworkProps;

  /**
   * The manager EC2 instance configuration. If not set, the defaults will be used.
   * @link GitlabRunnerAutoscalingManagerBaseProps
   */
  readonly manager?: GitlabRunnerAutoscalingManagerBaseProps;

  /**
   * The runner EC2 instances settings. At least one runner should be set up.
   * @link GitlabRunnerAutoscalingJobRunnerProps
   */
  readonly runners: GitlabRunnerAutoscalingJobRunnerProps[];
}

/**
 * The distributed GitLab runner S3 cache. Either pass an existing bucket or override default options.
 * @see https://docs.gitlab.com/runner/configuration/advanced-configuration.html#the-runnerscaches3-section
 */
export interface GitlabRunnerAutoscalingCacheProps {
  /**
   * An existing S3 bucket used as runner's cache.
   */
  readonly bucket?: IBucket;

  /**
   * If no existing S3 bucket is provided, a S3 bucket will be created.
   */
  readonly options?: CacheProps;
}

/**
 * The Gitlab Runner autoscaling on EC2 by Docker Machine.
 *
 * @example <caption>Provisioning a basic Runner</caption>
 * const app = new cdk.App();
 * const stack = new cdk.Stack(app, "RunnerStack", {
 *   env: {
 *     account: "000000000000",
 *     region: "us-east-1",
 *   }
 * });
 *
 * const token = new StringParameter(stack, "imported-token", {
 *   parameterName: "/gitlab-runner/token1",
 *   stringValue: gitlabToken,
 *   type: ParameterType.SECURE_STRING,
 *   tier: ParameterTier.STANDARD,
 * });
 *
 * new GitlabRunnerAutoscaling(stack, "GitlabRunner", {
 *   runners: [{
 *     token: "xxxxxxxxxxxxxxxxxxxx"
 *     }],
 * });
 */
export class GitlabRunnerAutoscaling extends Construct {
  readonly concurrent?: number;

  readonly checkInterval?: number;

  readonly logFormat?: LogFormat;

  readonly logLevel?: LogLevel;

  readonly network: Network;

  readonly cacheBucket: IBucket;

  readonly manager: GitlabRunnerAutoscalingManager;

  readonly runners: GitlabRunnerAutoscalingJobRunner[];

  constructor(scope: Stack, id: string, props: GitlabRunnerAutoscalingProps) {
    super(scope, id);
    const {
      concurrent,
      checkInterval,
      logFormat,
      logLevel,
      manager,
      cache,
      runners,
      network,
    }: GitlabRunnerAutoscalingProps = props;

    /**
     * Global Configuration
     * @link GlobalConfiguration
     */
    this.concurrent = concurrent ?? 10;
    this.checkInterval = checkInterval ?? 0;
    this.logFormat = logFormat ?? "runner";
    this.logLevel = logLevel ?? "info";

    /**
     * S3 Bucket for Runners' cache
     */
    this.cacheBucket = cache?.bucket || new Cache(this, "Cache", cache?.options).bucket;

    /**
     * Network
     */
    this.network = new Network(this, "Network", network);

    /**
     * Security groups
     */
    const runnersSecurityGroup = new SecurityGroup(this, "RunnersSecurityGroup", {
      description: "Security group for GitLab Runners.",
      vpc: this.network.vpc,
    });
    const managerSecurityGroup = new SecurityGroup(this, "ManagerSecurityGroup", {
      vpc: this.network.vpc,
      description: "Security group for GitLab Runners Manager.",
    });
    managerSecurityGroup.connections.allowTo(runnersSecurityGroup, Port.tcp(22), "SSH traffic from Manager");
    managerSecurityGroup.connections.allowTo(runnersSecurityGroup, Port.tcp(2376), "SSH traffic from Docker");

    this.runners = runners.map((runnerProps, index): GitlabRunnerAutoscalingJobRunner => {
      return new GitlabRunnerAutoscalingJobRunner(this, `GitlabRunnerAutoscalingJobRunner${index}`, runnerProps);
    });

    /**
     * GitLab Manager
     */
    this.manager = new GitlabRunnerAutoscalingManager(this, "Manager", {
      ...manager,
      globalConfiguration: {
        concurrent: this.concurrent,
        checkInterval: this.checkInterval,
        logFormat: this.logFormat,
        logLevel: this.logLevel,
      },
      runnersSecurityGroup: runnersSecurityGroup,
      network: this.network,
      cacheBucket: this.cacheBucket,
      runners: this.runners,
    });

    new AutoScalingGroup(this, "ManagerAutoscalingGroup", {
      vpc: this.network.vpc,
      vpcSubnets: {
        subnets: [this.network.subnet],
      },
      instanceType: this.manager.instanceType,
      machineImage: this.manager.machineImage,
      keyName: this.manager.keyPairName,
      securityGroup: managerSecurityGroup,
      role: this.manager.role,
      userData: this.manager.userData,
      init: this.manager.initConfig,
      initOptions: {
        ignoreFailures: false,
      },
      maxCapacity: 1,
      minCapacity: 1,
      desiredCapacity: 1,
      signals: Signals.waitForCount(1, { timeout: Duration.minutes(15) }),
      requireImdsv2: true,
    });
  }
}
