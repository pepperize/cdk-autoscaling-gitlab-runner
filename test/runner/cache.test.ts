import { App, Duration, Stack } from "aws-cdk-lib";
import { Capture, Template } from "aws-cdk-lib/assertions";
import { Cache as CacheConstruct, CacheProps as CacheConstructProps } from "../../src/runner/cache";

const stackProps = {
  env: {
    account: "123456789012",
    region: "us-east-1",
  },
};

describe("Cache", () => {
  it("Should match snapshot when Cache is being set", () => {
    // Given
    const app = new App();
    const stack = new Stack(app, "Stack", stackProps);

    // When
    new CacheConstruct(stack, "cache", {});
    const template = Template.fromStack(stack);

    // Then
    expect(template).toMatchSnapshot();
  });

  it("Should have cache expiration disabled when expiration is set to 0", () => {
    // Given
    const app = new App();
    const stack = new Stack(app, "a-very-very-long-stack-name", stackProps);
    const props: CacheConstructProps = {
      bucketName: "bucket",
      expiration: Duration.days(0),
    };

    // When

    new CacheConstruct(stack, "cache", props);
    const template = Template.fromStack(stack);

    // Then
    template.hasResourceProperties("AWS::S3::Bucket", {
      LifecycleConfiguration: {
        Rules: [
          {
            ExpirationInDays: 0,
            Status: "Disabled",
          },
        ],
      },
    });
  });

  it("Should substring last 63 characters of bucket name when provided a very long bucket name", () => {
    // Given
    const app = new App();
    const stack = new Stack(app, "a-very-very-long-stack-name", stackProps);
    const props: CacheConstructProps = {
      bucketName: "and-a-very-very-long-bucket-name",
      expiration: Duration.days(0),
    };
    new CacheConstruct(stack, "cache", props);

    // When
    const template = Template.fromStack(stack);

    // Then
    template.hasResourceProperties("AWS::S3::Bucket", {
      BucketName: "ck-name-and-a-very-very-long-bucket-name-123456789012-us-east-1",
    });
  });

  it("Should set ExpirationInDays to default (30) with the status Enabled when expiration is not set", () => {
    // Given
    const app = new App();
    const stack = new Stack(app, "stack", stackProps);
    const props: CacheConstructProps = {
      bucketName: "bucket",
    };
    new CacheConstruct(stack, "cache", props);

    // When
    const template = Template.fromStack(stack);

    // Then
    template.hasResourceProperties("AWS::S3::Bucket", {
      LifecycleConfiguration: {
        Rules: [
          {
            ExpirationInDays: 30,
            Status: "Enabled",
          },
        ],
      },
    });
  });

  it("Should set ExpirationInDays to 1 with the status Enabled when expiration is set to 1", () => {
    // Given
    const app = new App();
    const stack = new Stack(app, "stack", stackProps);
    const props: CacheConstructProps = {
      bucketName: "bucket",
      expiration: Duration.days(1),
    };
    new CacheConstruct(stack, "cache", props);

    // When
    const template = Template.fromStack(stack);
    const capture = new Capture();
    template.hasResourceProperties("AWS::S3::Bucket", {
      LifecycleConfiguration: capture,
    });

    // Then
    expect(capture.asObject()).toEqual({
      Rules: [
        {
          ExpirationInDays: 1,
          Status: "Enabled",
        },
      ],
    });
  });
});
