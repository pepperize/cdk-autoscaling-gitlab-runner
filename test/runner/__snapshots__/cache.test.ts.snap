// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cache Should match snapshot when Cache is being set 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "CacheBucket41D9D0B0": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "BucketEncryption": Object {
          "ServerSideEncryptionConfiguration": Array [
            Object {
              "ServerSideEncryptionByDefault": Object {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
        "BucketName": "stack-runner-cache-123456789012-us-east-1",
        "LifecycleConfiguration": Object {
          "Rules": Array [
            Object {
              "ExpirationInDays": 30,
              "Status": "Enabled",
            },
          ],
        },
        "PublicAccessBlockConfiguration": Object {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "Tags": Array [
          Object {
            "Key": "aws-cdk:auto-delete-objects",
            "Value": "true",
          },
        ],
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Delete",
    },
    "CacheBucketAutoDeleteObjectsCustomResource12E72BBC": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "CacheBucketPolicyE2F00ED4",
      ],
      "Properties": Object {
        "BucketName": Object {
          "Ref": "CacheBucket41D9D0B0",
        },
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "CustomS3AutoDeleteObjectsCustomResourceProviderHandler9D90184F",
            "Arn",
          ],
        },
      },
      "Type": "Custom::S3AutoDeleteObjects",
      "UpdateReplacePolicy": "Delete",
    },
    "CacheBucketPolicyE2F00ED4": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "CacheBucket41D9D0B0",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:*",
              "Condition": Object {
                "Bool": Object {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": Object {
                "AWS": "*",
              },
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "CacheBucket41D9D0B0",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "CacheBucket41D9D0B0",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": Array [
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
              ],
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::GetAtt": Array [
                    "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
                    "Arn",
                  ],
                },
              },
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "CacheBucket41D9D0B0",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "CacheBucket41D9D0B0",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "CustomS3AutoDeleteObjectsCustomResourceProviderHandler9D90184F": Object {
      "DependsOn": Array [
        "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": "cdk-hnb659fds-assets-123456789012-us-east-1",
          "S3Key": "bdb08b9984816ecf43301a0c6d8913e270ee2448940b6b95e430dd222a4d90be.zip",
        },
        "Description": Object {
          "Fn::Join": Array [
            "",
            Array [
              "Lambda function for auto-deleting objects in ",
              Object {
                "Ref": "CacheBucket41D9D0B0",
              },
              " S3 bucket.",
            ],
          ],
        },
        "Handler": "__entrypoint__.handler",
        "MemorySize": 128,
        "Role": Object {
          "Fn::GetAtt": Array [
            "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Sub": "arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
